aws-cpp-sdk-transfer/source/transfer/UploadFileRequest.cpp

```diff
@@ -779,9 +751,17 @@ bool UploadFileRequest::ProcessAvailableBuffers()
 
     std::shared_ptr<UploadBuffer> thisBuffer;
 
-    while (!DoneWithRequests() && GetReadyBuffer(thisBuffer))
+    while (!DoneWithRequests() && !IsCancelled())
     {
-        ProcessBuffer(thisBuffer);
+        if (GetReadyBuffer(thisBuffer))
+        {
+            ProcessBuffer(thisBuffer);
+        }
+        else
+        {
+            // TODO: sleep/wait instead?

```

Comment:
RESOLUTION: all this should disappear. There should be an array (or list) of buffers, a latch and a wait-post (conditional variable) that people wait on.
We may get around to this next week.